### PR TITLE
KTOR-4771 Fix Receive Non-Nullable null Message

### PR DESCRIPTION
**Pull request by e5l**
Fix [KTOR-4771](https://youtrack.jetbrains.com/issue/KTOR-4771)

### CODE REVIEW

File: ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/request/ApplicationReceiveFunctions.kt [hunk 1]
```diff
@@ -58,7 +58,8 @@ public suspend inline fun <reified T : Any> ApplicationCall.receiveOrNull(): T?
  * @return instance of [T] received from this call.
  * @throws ContentTransformationException when content cannot be transformed to the requested type.
  */
-public suspend inline fun <reified T : Any> ApplicationCall.receive(): T = receiveNullable(typeInfo<T>())!!
+public suspend inline fun <reified T : Any> ApplicationCall.receive(): T = receiveNullable(typeInfo<T>())
+    ?: throw CannotTransformContentToTypeException(typeInfo<T>().kotlinType!!)
```
Comment (Stexxe) [inline ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/request/ApplicationReceiveFunctions.kt:62]: Also `CannotTransformContentToTypeException` will be thrown if transformation result is `null` and it's successful.

Comment (Stexxe) [reply ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/request/ApplicationReceiveFunctions.kt:62]: Why do need to throw the same exception when it's raised already in `ApplicationCall.receiveNullable`? The only case it covers is when a result is `NullBody` so the exception is misleading.

File: ktor-server/ktor-server-core/jvmAndNix/test/io/ktor/server/application/ApplicationCallReceiveTest.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                    NullBody
+                }
+            }
+        }
+
+        install(nullPlugin)
+
+        routing {
+            post("/") {
+                val result: String = try {
```
Comment (Stexxe) [inline ktor-server/ktor-server-core/jvmAndNix/test/io/ktor/server/application/ApplicationCallReceiveTest.kt:34]: Could be simplified with the `assertFailsWith` inside route's hanlder

File: ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/request/ApplicationReceiveFunctions.kt (hunk 1 above)
Comment (e5l) [reply ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/request/ApplicationReceiveFunctions.kt:62]: We throw `ContentTransformationException` if content cannot be transformed to the requested type. If content is converted to null for non nullable type it's `ContentTransformationException`

Comment (rsinukov) [inline ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/request/ApplicationReceiveFunctions.kt:62]: maybe extract `typeInfo<T>()` to val?

**Review by rsinukov — APPROVED** [review]

File: ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/request/ApplicationReceiveFunctions.kt (hunk 1 above)
Comment (e5l) [reply ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/request/ApplicationReceiveFunctions.kt:62]: I would keep it as one-liner cause `typeInfo` is compile time constant

**Pull request merged**